### ISSUE TITLE
LiveDataK wrapper for arrow-effects-android

### ISSUE THREAD

**Issue by JorgeCastilloPrz** [arrow-kt/arrow#1095]
Let's start working on `arrow-effects-android` adding a `LiveData` wrapper (`LiveDataK`). Create the arrow module and add the wrapper along with the required type class instances and tests for it.

**Comment by JorgeCastilloPrz** [comment]
@Tylos will take this one. It's a good one for an Android dev to start getting in contact with the key arrow concepts (data types, typeclasses, instances of the type classes for the data type, testing laws over the data type).

**Comment by raulraja** [comment]
I think this would require `Stream<F, A>` to be properly modeled and if we want to include the different lifecycle methods in a similar fashion as to how it's done with Rx2 traditionally in Android. The difference is that our Stream works for Deferred, Observable, IO, etc.

**Comment by JorgeCastilloPrz** [comment]
@raulraja I think `LiveData` does not model anything in terms of lifecycle. It's an `Observable`. It just queries the observer before emitting (which must implement  `LifecycleOwner` or somethjing similar), like `observer.isResumed()`, `observer.isStarted()`and so on. So I'm not sure it requires modeling anything related to the lifecycle as a stream. We need to investigate a bit more.

**Comment by pakoito** [comment]
https://github.com/arrow-kt/arrow/pull/475

I did it already. tl;dr it's so slow it takes literal minutes to do tailrec. Unusable.

**Comment by JorgeCastilloPrz** [comment]
Can you ellaborate on how to proceed then? Is it a problem on how `LiveData.switchMap` is coded? can't we do any workarounds to have this wrapper in place?

**Comment by pakoito** [comment]
It's coded so it does excessive checking and they don't optimize composition. It's all intentional so not much we can do about it other than ask Google. I already did, and their response was this this constructs are not meant to be used a lot to chain them.

**Comment by JorgeCastilloPrz** [comment]
Ok so, given that this new data type we potentially work on would be "equivalent" to `LiveData` in terms of public APIs, **but efficient**, I assume we would want it to work with Google's LifecycleOwner observers?. SDK activities and fragments are already LifecycleOwners so if we can work with that interface we'd not need to reencode a lot of things and create our own Activity / Fragment parent classes (which is potentially problematic because of the classic enforced inheritance problems).

**Comment by JorgeCastilloPrz** [comment]
I see. Sounds good. Is android-ui blocked by arrow-streams then? Looking for ways to put some efforts on this for Q1, but not sure it's gonna be possible.

**Comment by pakoito** [comment]
@Cotel can fill you in better!
